packages/babel-plugin-transform-decorators-2/src/index.js

```diff
@@ -0,0 +1,183 @@
+      if (!method.isClassMethod()) continue;
+      if (!node.decorators || !method.node.decorators.length) continue;
+
+      const decorators = method.node.decorators.map(d => d.expression);
+
+      node.decorators = []; //TODO: should we remove using path methods? method.get("decorators") doesn't work
+
+      const entry = [getKey(node), t.arrayExpression(decorators)];
+
+      if (node.static) {
+        entry.push(t.booleanLiteral(true));
+      }
+
+      entries.push(t.arrayExpression(entry));
+    }
+
+    return t.arrayExpression(entries);
+  }
+
+  // expects path of a ClassExpression
+  function takeClassDecorators(path) {
+    if (path.node.decorators && path.node.decorators.length) {
+      const decorators = path.node.decorators.map(d => d.expression);
+      path.node.decorators = [];
+      return t.arrayExpression(decorators);
+    } else {
+      return t.arrayExpression();
+    }
+  }
+
+  // assume:addKeysToComputedMembers has been executed
+  function undecoratedMethods(path) {
+    const body = path.get("body.body");
+    const result = []; // shape: [["method1"], ["method2"], ["staticMethod1", true]]
+
+    for (const method of body) {
+      if (method.node.decorators && method.node.decorators.length > 0) continue;
+      if (method.node.key.name == "constructor") continue;
+
+      const key = getKey(method.node);
+
+      if (method.node.static) {
+        result.push(t.arrayExpression([key, t.booleanLiteral(true)]));
+      } else {
+        result.push(t.arrayExpression([key]));
+      }
+    }
+
+    return t.arrayExpression(result);
+  }
+
+  return {
+    inherits: syntaxDecorators2,
+    visitor: {
+      // export default is a special case since it expects and expression rather than a declaration
+      ExportDefaultDeclaration(path) {
+        const classPath = path.get("declaration");
+        if (!classPath.isClassDeclaration()) return;
+
+        if (classPath.node.id) {
+          const ref = classPath.node.id;
+          path.insertBefore(
+            t.variableDeclaration("let", [t.variableDeclarator(ref)]),
+          );
+          classPath.replaceWith(
+            t.assignmentExpression("=", ref, t.toExpression(classPath.node)),
+          );
+        } else {
+          classPath.replaceWith(t.toExpression(classPath));
+        }
+      },
+
+      // replace declaration with a let declaration
+      ClassDeclaration(path) {
+        const { node } = path;
+        const ref = node.id || path.scope.generateUidIdentifier("class");
+
+        path.replaceWith(
+          t.variableDeclaration("let", [
+            t.variableDeclarator(ref, t.toExpression(node)),
+          ]),
+        );
+      },
+
+      ClassExpression(path, file) {
+        file.addHelper("makeElementDescriptor");
+        file.addHelper("mergeDuplicateElements");
+        file.addHelper("decorateClass", ["mergeDuplicateElements"]);
+        file.addHelper("decorateElement", ["mergeDuplicateElements"]);
```

Comment:
It would be better to include these helpers only if the class has some decorators.